browser/modules/PlacesProvider.jsm

```diff
@@ -0,0 +1,258 @@
+ */
+let Links = {
+  /**
+   * EventEmitter interface
+   */
+  eventEmitter: new EventEmitter(),
+  on: (...params) => this.eventEmitter.on(...params),
+  once: (...params) => this.eventEmitter.once(...params),
+  off: (...params) => this.eventEmitter.off(...params),
+
+  /**
+   * Set this to change the maximum number of links the provider will provide.
+   */
+  maxNumLinks: HISTORY_RESULTS_LIMIT,
+
+  /**
+   * A set of functions called by @mozilla.org/browser/nav-historyservice
+   * All history events are emitted from this object.
+   */
+  historyObserver: {
+    onDeleteURI: function PlacesProvider_onDeleteURI(aURI) {
+      // let observers remove sensetive data associated with deleted visit
+      Links.eventEmitter.emit("deleteURI", {
+        url: aURI.spec,
+      });
+    },
+
+    onClearHistory: function() {
+      Links.eventEmitter.emit("clearHistory");
+    },
+
+    onFrecencyChanged: function PlacesProvider_onFrecencyChanged(aURI,
+                           aNewFrecency, aGUID, aHidden, aLastVisitDate) { // jshint ignore:line
+      // The implementation of the query in getLinks excludes hidden and
+      // unvisited pages, so it's important to exclude them here, too.
+      if (!aHidden && aLastVisitDate) {
+        Links.eventEmitter.emit("linkChanged", {
+          url: aURI.spec,
+          frecency: aNewFrecency,
+          lastVisitDate: aLastVisitDate,
+          type: "history",
+        });
+      }
+    },
+
+    onManyFrecenciesChanged: function PlacesProvider_onManyFrecenciesChanged() {
+      Links.eventEmitter.emit("manyLinksChanged");
+    },
+
+    onTitleChanged: function PlacesProvider_onTitleChanged(aURI, aNewTitle) {
+      Links.eventEmitter.emit("linkChanged", {
+        url: aURI.spec,
+        title: aNewTitle
+      });
+    },
+
+    QueryInterface: XPCOMUtils.generateQI([Ci.nsINavHistoryObserver,
+                        Ci.nsISupportsWeakReference])
+  },
+
+  /**
+   * Must be called before the provider is used.
+   */
+  init: function PlacesProvider_init() {
+    PlacesUtils.history.addObserver(this.historyObserver, true);
+  },
+
+  /**
+   * Gets the current set of links delivered by this provider.
+   *
+   * @returns {Promise} Returns a promise with the array of links as payload.
+   */
+  getLinks: function PlacesProvider_getLinks() {
+    let getLinksPromise = new Promise((resolve, reject) => {
+      let options = PlacesUtils.history.getNewQueryOptions();
+      options.maxResults = this.maxNumLinks;
+
+      // Sort by frecency, descending.
+      options.sortingMode = Ci.nsINavHistoryQueryOptions
+        .SORT_BY_FRECENCY_DESCENDING;
+
+      let links = [];
+
+      let callback = {
+        handleResult: function(aResultSet) {
+          let row;
+
+          while ((row = aResultSet.getNextRow())) {
+            let url = row.getResultByIndex(1);
+            if (LinkChecker.checkLoadURI(url)) {
+              let title = row.getResultByIndex(2);
```

Comment:
We should just put these into object. 
```JS
              links.push({
                url,
                title: row.getResultByIndex(2),
                frecency: row.getResultByIndex(12);,
                lastVisitDate: row.getResultByIndex(5),
                type: "history",
              });
```

Also, you might want to make the link into a variable before `push`ing it. Just makes it a bit easier to debug later if there are problems.